designs/2021-stricter-rule-test-validation/README.md

```diff
@@ -0,0 +1,279 @@
+
+This continues a trend in which `RuleTester` validation has become stricter with each major release of ESLint.
+
+## Detailed Design
+
+<!--
+   This is the bulk of the RFC.
+
+   Explain the design with enough detail that someone familiar with ESLint
+   can implement it by reading this document. Please get into specifics
+   of your approach, corner cases, and examples of how the change will be
+   used. Be sure to define any new terms in this section.
+-->
+
+This section contains a sub-section for each new category of assertion.
+
+Code changes:
+
+- Assertions will be added in the `lib/rule-tester/rule-tester.js` file in the `testInvalidTemplate()` function
+- Tests will be added for each assertion in `tests/lib/rule-tester/rule-tester.js`
+
+### Invalid test case error objects must contain `suggestions` if the test case produces suggestions
+
+This would enforce:
+
+```pt
+AssertionError [ERR_ASSERTION]: The rule produced suggestion(s). Please add a 'suggestions' property.
+```
+
+This change has precedent in the existing behavior today where invalid test cases that produce an autofix require an `output` assertion (this behavior was added as a [breaking change](https://eslint.org/docs/user-guide/migrating-to-7.0.0#additional-validation-added-to-the-ruletester-class) in ESLint v7):
+
+```pt
+AssertionError [ERR_ASSERTION]: The rule fixed the code. Please add 'output' property.
+```
+
+More about motivation:
+
+- It's easy to forget to assert what suggestions a test case produces, or that a test case produces no assertions
+- It's inconvenient to have to add `suggestions: []` to every single invalid test case that does not produce suggestions, in order to assert that they don't produce suggestions
+- Suggestions are a newer feature, and many existing test cases may not have been updated to test newly-added suggestions
+
+### Invalid test case error and suggestion objects must include the message
+
+This would enforce:
+
+```pt
+Test error object must contain exactly one of 'message' or 'messageId'.
+```
+
+```pt
+Test suggestion object must contain exactly one of 'messageId' or 'desc'.
+```
+
+This has the effect of enforcing that error and suggestion objects cannot be empty, which is regrettably allowed today. Note that if we decided that the message should not be required in test cases, we would instead want to add an assertion that test case objects are not empty at the minimum.
+
+Note that there's an existing assertion that already enforces part of this:
+
+```pt
+AssertionError [ERR_ASSERTION]: Error should not specify both 'message' and a 'messageId'.
+```
+
+### Each suggestion object requires the `output` property
+
+In addition to the message provided by the suggestion, the actual code output is another key part of the suggestion that tests should require:
+
+```pt
+AssertionError [ERR_ASSERTION]: Test suggestion object is missing 'output`.
+```
+
+### Disallow repeating the test case `code` as `output` for non-fixable test cases
```

Comment:
```suggestion
### Disallow repeating the test case `code` as `output`
```

Should we broaden this and also apply it to suggestions to assert that fixes and suggestions actually change the code? For example, `fix: fixer => fixer.insertTextBefore(node, "")` is a no-op.

Reply:
Agreed with this, updated to say that both autofixes and suggestions must change the code.